x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/Migration.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.profiler;
+
+import org.elasticsearch.action.admin.indices.mapping.put.PutMappingRequest;
+import org.elasticsearch.action.admin.indices.settings.put.UpdateSettingsRequest;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * Encapsulates a single migration action.
+ */
+public interface Migration {
+    /**
+     * @return The index template version that the targeted index will be on after the migration has been applied.
+     */
+    int getTargetIndexTemplateVersion();
+
+    /**
+     * Applies this migration. Depending on the type of migration one or the other consumer will be called.
+     *
+     * @param index The name of the specific index to which the migration should be applied.
+     * @param mappingConsumer A consumer that applies mapping changes.
+     * @param settingsConsumer A consumer that applies settings changes.
+     */
+    void apply(String index, Consumer<PutMappingRequest> mappingConsumer, Consumer<UpdateSettingsRequest> settingsConsumer);
+
+    /**
+     * Builds migrations for an index.
+     */
+    class Builder {
+        private final List<Migration> migrations = new ArrayList<>();
+        private Integer targetVersion;
+
+        private void checkVersionSet() {
+            if (targetVersion == null) {
+                throw new IllegalStateException("Set targetVersion before defining migrations");
+            }
+        }
+
+        /**
+         * @param version The index template version that the targeted index will reach after this migration has been applied.
+         * @return <code>this</code> to allow for method chaining.
+         */
+        public Builder migrateToIndexTemplateVersion(int version) {
```

Comment:
Just for my curiosity, why not have a constructor taking the version to avoid the `checkVersionSet()` function?

Reply:
We could probably do that for the initial version but we allow for migrations across multiple versions and thus we need to be able to call it multiple times. But there's definitely multiple ways to approach this. :)